### PR TITLE
refactor(compiler): replace `Comment` nodes with `leadingComments` property

### PR DESCRIPTION
**Pull request by petebacondarwin**
Common AST formats such as TS and Babel do not use a separate
node for comments, but instead attach comments to other AST nodes.
Previously this was worked around in TS by creating a `NotEmittedStatement`
AST node to attach the comment to. But Babel does not have this facility,
so it will not be a viable approach for the linker.

This commit refactors the output AST, to remove the `CommentStmt` and
`JSDocCommentStmt` nodes. Instead statements have a collection of
`leadingComments` that are rendered/attached to the final AST nodes
when being translated or printed.

### CODE REVIEW

File: packages/compiler-cli/src/transformers/node_emitter.ts [hunk 1]
```diff
@@ -292,6 +281,9 @@ export class NodeEmitterVisitor implements StatementVisitor, ExpressionVisitor {
     if (tsNode && !this._nodeMap.has(tsNode)) {
       this._nodeMap.set(tsNode, ngNode);
     }
+    if (tsNode !== null && ngNode instanceof Statement) {
+      attachComments(tsNode as unknown as ts.Statement, ngNode.leadingComments);
+    }
```
Comment (JoostK) [inline packages/compiler-cli/src/transformers/node_emitter.ts:286]: I think that adding this here prompts for the method to be renamed as well, maybe `postprocess` or something?

File: packages/compiler-cli/test/transformers/node_emitter_spec.ts [hunk 2]
```diff
@@ -273,42 +273,47 @@ describe('TypeScriptNodeEmitter', () => {
     });
 
     it('should support singleline comments', () => {
-      expect(emitStmt(new o.CommentStmt('Simple comment'))).toBe('// Simple comment');
+      expect(emitStmt(new o.ReturnStatement(o.literal(1), null, [o.leadingComment('a\nb', false)])))
+          .toBe('// a\n// a\n return 1;');
```
Comment (JoostK) [inline packages/compiler-cli/test/transformers/node_emitter_spec.ts:277]: Where did the `'b'` go on line 2, it's turned into an `'a'` here :-)

**Review by JoostK — APPROVED** [review]
LGTM but CI is unhappy.

**Comment by nutza249949** [comment]
- ~~
### 
## ~~

File: packages/compiler/src/output/abstract_emitter.ts [hunk 3]
```diff
@@ -248,25 +267,12 @@ export abstract class AbstractEmitterVisitor implements o.StatementVisitor, o.Ex
   abstract visitTryCatchStmt(stmt: o.TryCatchStmt, ctx: EmitterVisitorContext): any;
 
   visitThrowStmt(stmt: o.ThrowStmt, ctx: EmitterVisitorContext): any {
+    this.printLeadingComments(stmt, ctx);
```
Comment (gkalpak) [inline packages/compiler/src/output/abstract_emitter.ts:270]: Why do some `visit*` methods have this line and others don't? How do we determine which methods need it?

File: packages/compiler/src/output/output_ast.ts [hunk 4]
```diff
@@ -16,15 +16,11 @@ export enum TypeModifier {
 }
 
 export abstract class Type {
-  constructor(public modifiers: TypeModifier[]|null = null) {
-    if (!modifiers) {
-      this.modifiers = [];
-    }
-  }
+  constructor(public modifiers: TypeModifier[] = []) {}
```
Comment (gkalpak) [inline packages/compiler/src/output/output_ast.ts:19]: Why this change here (and below)? Won't it result into many unnecessary arrays being instantiated?

File: packages/compiler-cli/src/ngtsc/modulewithproviders/src/scanner.ts [hunk 5]
```diff
@@ -103,7 +103,7 @@ export class ModuleWithProvidersScanner {
         this.emitter.emit(ngModule, decl.getSourceFile(), ImportFlags.ForceNewImport);
     const ngModuleType = new ExpressionType(ngModuleExpr);
     const mwpNgType = new ExpressionType(
-        new ExternalExpr(Identifiers.ModuleWithProviders), /* modifiers */ null, [ngModuleType]);
+        new ExternalExpr(Identifiers.ModuleWithProviders), /* modifiers */[], [ngModuleType]);
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/modulewithproviders/src/scanner.ts:106]: Talk to the automatic formatter!

File: packages/compiler/src/output/abstract_emitter.ts (hunk 3 above)
Comment (petebacondarwin) [reply packages/compiler/src/output/abstract_emitter.ts:270]: Only methods that visit statements need this since we are not supporting adding comments within a statement.

File: packages/compiler/src/output/abstract_emitter.ts [hunk 6]
```diff
@@ -202,13 +202,31 @@ export class EmitterVisitorContext {
 export abstract class AbstractEmitterVisitor implements o.StatementVisitor, o.ExpressionVisitor {
   constructor(private _escapeDollarInStrings: boolean) {}
 
+  protected printLeadingComments(stmt: o.Statement, ctx: EmitterVisitorContext): void {
+    for (const comment of stmt.leadingComments) {
+      if (comment instanceof o.JSDocComment) {
+        ctx.print(stmt, `/*${comment.toString()}*/`, comment.trailingNewline);
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler/src/output/abstract_emitter.ts:208]: The `JSDocComment.toString()` function returns a string starting with `*`.

File: packages/compiler-cli/src/transformers/node_emitter.ts [hunk 7]
```diff
@@ -31,30 +31,19 @@ export class TypeScriptNodeEmitter {
...
-    const preambleStmts: ts.Statement[] = [];
+    const sourceStatements =
+        [...converter.getReexports(), ...converter.getImports(), ...statements];
     if (preamble) {
-      const commentStmt = this.createCommentStatement(sourceFile, preamble);
-      preambleStmts.push(commentStmt);
+      if (statements.length === 0) {
+        statements.push(ts.createNotEmittedStatement(sourceFile));
+      }
+      attachComments(statements[0], [new LeadingComment(preamble, true)]);
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/transformers/node_emitter.ts]: Yes! Thank you so much @gkalpak - I totally missed this.

File: packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts [hunk 8]
```diff
@@ -1663,12 +1663,12 @@ runInEachFileSystem(() => {
...
         mainNgcc({basePath: '/dist', propertiesToConsider: ['es2015']});
         const jsContents = fs.readFile(_(`/dist/local-package/index.js`));
-        expect(jsContents).toContain('/** @nocollapse */ \nAppComponent.ɵcmp =');
+        expect(jsContents).toContain('/** @nocollapse */\nAppComponent.ɵcmp =');
       });
       it('should default to not give closure annotated output', () => {
         mainNgcc({basePath: '/dist', propertiesToConsider: ['es2015']});
         const jsContents = fs.readFile(_(`/dist/local-package/index.js`));
-        expect(jsContents).not.toContain('/** @nocollapse */');
+        expect(jsContents).not.toContain('@nocollapse');
```
Comment (petebacondarwin) [inline packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:1671]: This change just makes it more resilient to the comment structure.

File: packages/compiler/src/output/output_ast.ts (hunk 4 above)
Comment (petebacondarwin) [reply packages/compiler/src/output/output_ast.ts:19]: I changed the derived classes below to pass `undefined`, so this is the only place where a `[]` is created now.

File: packages/compiler-cli/src/transformers/node_emitter.ts [hunk 9]
```diff
@@ -773,6 +746,20 @@ export class NodeEmitterVisitor implements StatementVisitor, ExpressionVisitor {
...
+    for (const piece of pieces) {
+      ts.addSyntheticLeadingComment(
+          statement,
+          comment.multiline ? ts.SyntaxKind.MultiLineCommentTrivia :
+                              ts.SyntaxKind.SingleLineCommentTrivia,
+          piece, /** hasTrailingNewLine */ !comment.multiline || comment.trailingNewline);
+    }
+  }
+  return statement;
+}
```
Comment (gkalpak) [inline packages/compiler-cli/src/transformers/node_emitter.ts:762]: OOC, why not use the same implementation as in [ngtsc/translator/src/translator.ts#L423-L438][1] (and possibly share a helper function)?

[1]: https://github.com/angular/angular/blob/2ff26f32d126107ce7dbb4ba63ec00a94ef8fafd/packages/compiler-cli/src/ngtsc/translator/src/translator.ts#L423-L438

**Review by gkalpak — APPROVED** [review]
LGTM :+1:

File: packages/compiler/src/output/output_ast.ts (hunk 4 above)
Comment (gkalpak) [reply packages/compiler/src/output/output_ast.ts:19]: This makes the code cleaner, but we will still get all those empty arrays eventually created, right?
I guess it is not a concern, but if this class is instantiated a lot, it might be worth checking some numbers (just in case).

Comment (petebacondarwin) [reply packages/compiler/src/output/output_ast.ts:19]: Hmmm good point.

File: packages/compiler-cli/src/transformers/node_emitter.ts (hunk 9 above)
Comment (petebacondarwin) [reply packages/compiler-cli/src/transformers/node_emitter.ts:762]: Yeah I thought about it, but I was concerned that one is ViewEngine the other Ivy. And I wasn't sure where such a shared util would live...

File: packages/compiler/src/output/output_ast.ts (hunk 4 above)
Comment (petebacondarwin) [reply packages/compiler/src/output/output_ast.ts:19]: Actually look down at the previous code - line 21. It was converting `null` to an array anyway.

Comment (JoostK) [reply packages/compiler/src/output/output_ast.ts:19]: I was also concerned about memory usage here, but figured that the code in the constructor would have always initialized it with an empty array regardless:

```ts
    if (!modifiers) {		
       this.modifiers = [];		
     }
```
(that was removed in this PR).

That's also why `this.modifiers!` was used in a couple places.

Come to think of it, this PR probably does introduce allocations of arrays for the comments, that might benefit from being lazily initialized. I meant to mention that in my first round of comments but forgot about it.

File: packages/compiler-cli/src/transformers/node_emitter.ts (hunk 9 above)
Comment (petebacondarwin) [reply packages/compiler-cli/src/transformers/node_emitter.ts]: Actual I have given this a go. Let's see if CI breaks.

File: packages/compiler/src/output/output_ast.ts (hunk 4 above)
Comment (petebacondarwin) [reply packages/compiler/src/output/output_ast.ts:19]: Let me look into the comments arrays...

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:331837764:BASE:331837821:1600200661168:32e869b9).

File: packages/compiler/src/output/output_ast.ts [hunk 10]
```diff
@@ -994,17 +1001,22 @@ export abstract class Statement {
   abstract visitStatement(visitor: StatementVisitor, context: any): any;
 
   hasModifier(modifier: StmtModifier): boolean {
-    return this.modifiers!.indexOf(modifier) !== -1;
+    return this.modifiers.indexOf(modifier) !== -1;
+  }
+
+  addLeadingComment(leadingComment: LeadingComment): void {
+    this.leadingComments = this.leadingComments || [];
+    this.leadingComments.push(leadingComment);
```
Comment (JoostK) [inline packages/compiler/src/output/output_ast.ts:1009]: For future fancyness and making the code extra hard to read, this could be

```suggestion
    (this.leadingComments ??= []).push(leadingComment);
```

But please don't 😱 (probably won't even fly in g3, as I believe this is TS 4.0 and g3 might still be on 3.9?)

But at least it could be

```suggestion
    this.leadingComments = this.leadingComments ?? [];
    this.leadingComments.push(leadingComment);
```

File: packages/compiler-cli/src/ngtsc/translator/src/translator.ts [hunk 11]
```diff
@@ -424,6 +420,11 @@ class ExpressionTranslatorVisitor implements ExpressionVisitor, StatementVisitor
     return ts.createTypeOf(ast.expr.visitExpression(this, context));
   }
 
+  private attachComments<T extends ts.Statement>(statement: T, leadingComments?: LeadingComment[]):
+      T {
+    return attachComments(statement, leadingComments);
```
Comment (gkalpak) [inline packages/compiler-cli/src/ngtsc/translator/src/translator.ts:425]: OOC, why not use the function directly (instead of wrapping it in a method)? 😕

**Review by JoostK — APPROVED** [review]

**Review by gkalpak — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/translator/src/translator.ts (hunk 11 above)
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/translator/src/translator.ts:425]: Good point. It's previous incarnation used `this`. But this doesn't and could be a used directly.
My next PR (https://github.com/angular/angular/pull/38775) moves it out of the translator into the `AstFactory` interface, where it does need to be a method.

Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/translator/src/translator.ts:425]: Changed

File: packages/compiler/src/output/output_ast.ts (hunk 10 above)
Comment (petebacondarwin) [reply packages/compiler/src/output/output_ast.ts]: Done

**Comment by AndrewKushnir** [comment]
FYI, there are some failures in g3 with the changes from this PR. The problems are related to the file headers (block of comments at the very top of the file) processed by Closure. More investigation is needed, adding the "blocked" label for now...

**Comment by AndrewKushnir** [comment]
[Presubmit #2](http://test/OCL:332252480:BASE:332252504:1600360763501:74f3c4f7).
